File: Arcadia_v2/Creatures/AnimalGrowthCatalog.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;

namespace Arcadia_v2.Creatures
{
    public sealed record AnimalGrowthOption(int PartyIndex, Animal CurrentAnimal, Animal AdultAnimal);

    public static class AnimalGrowthCatalog
    {
        private static readonly Dictionary<string, string> AdultSpeciesByBaseSpecies = new()
        {
            ["CAT"] = "LION",
            ["DOG"] = "WOLF",
            ["HORSE"] = "STALLION",
            ["TURTLE"] = "TORTOISE",
            ["BIRD"] = "EAGLE",
            ["ANT"] = "BEE",
            ["CUB"] = "BEAR",
            ["SERPENT"] = "DRAGON"
        };

        public static IReadOnlyList<AnimalGrowthOption> GetGrowthOptions(Player player)
        {
            ArgumentNullException.ThrowIfNull(player);

            List<AnimalGrowthOption> options = new();

            for (int i = 0; i < player.AnimalInventory.Count; ++i)
            {
                Animal animal = player.AnimalInventory[i];

                if (player.GetBond(animal.Element) < 100)
                {
                    continue;
                }

                if (TryGetAdultForm(animal, out Animal? adultAnimal))
                {
                    options.Add(new AnimalGrowthOption(i, animal, adultAnimal!));
                }
            }

            return options;
        }

        public static bool HasGrowthOptions(Player player)
        {
            return GetGrowthOptions(player).Count > 0;
        }

        private static bool TryGetAdultForm(Animal animal, out Animal? adultAnimal)
        {
            string[] nameParts = animal.Name.Split('_', 2);

            if (nameParts.Length != 2 ||
                !AdultSpeciesByBaseSpecies.TryGetValue(nameParts[1], out string? adultSpecies))
            {
                adultAnimal = null;
                return false;
            }

            string adultName = $"{nameParts[0]}_{adultSpecies}";
            adultAnimal = AnimalFactory.CreateAnimals()
                .Single(candidate => candidate.Name == adultName)
                .Clone();

            return true;
        }
    }
}

File: arcadia_flutter/lib/creatures/animal_growth_catalog.dart
import '../player/player.dart';
import 'animal.dart';
import 'animal_catalog.dart';

class AnimalGrowthOption {
  const AnimalGrowthOption({
    required this.partyIndex,
    required this.currentAnimal,
    required this.adultAnimal,
  });

  final int partyIndex;
  final Animal currentAnimal;
  final Animal adultAnimal;
}

class AnimalGrowthCatalog {
  const AnimalGrowthCatalog._();

  static const Map<String, String> _adultSpeciesByBaseSpecies = {
    'CAT': 'LION',
    'DOG': 'WOLF',
    'HORSE': 'STALLION',
    'TURTLE': 'TORTOISE',
    'BIRD': 'EAGLE',
    'ANT': 'BEE',
    'CUB': 'BEAR',
    'SERPENT': 'DRAGON',
  };

  static List<AnimalGrowthOption> getGrowthOptions(Player player) {
    final options = <AnimalGrowthOption>[];

    for (var i = 0; i < player.animalInventory.length; i += 1) {
      final animal = player.animalInventory[i];

      if (player.getBond(animal.element) < 100) {
        continue;
      }

      final adultAnimal = tryCreateAdultForm(animal);
      if (adultAnimal != null) {
        options.add(
          AnimalGrowthOption(
            partyIndex: i,
            currentAnimal: animal,
            adultAnimal: adultAnimal,
          ),
        );
      }
    }

    return options;
  }

  static bool hasGrowthOptions(Player player) {
    return getGrowthOptions(player).isNotEmpty;
  }

  static Animal? tryCreateAdultForm(Animal animal) {
    final nameParts = animal.name.split('_');
    if (nameParts.length != 2) {
      return null;
    }

    final adultSpecies = _adultSpeciesByBaseSpecies[nameParts[1]];
    if (adultSpecies == null) {
      return null;
    }

    final adultName = '${nameParts[0]}_$adultSpecies';
    return AnimalCatalog.createAnimals()
        .singleWhere((candidate) => candidate.name == adultName)
        .clone();
  }
}
